File: 0078-subsets/0078-subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def backtrack_comb(start, arr, mx_len) :
            if len(arr) == mx_len :
                result.append(arr)
                return
            for idx in range(start, len(nums)) :
                backtrack_comb(idx+1, arr+[nums[idx]], mx_len)

        result = []
        for k in range(len(nums)+1) :
            backtrack_comb(0, [], k)    
        return result

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void backtrack_comb(int start, int mx_len, const vector<int>& nums, vector<vector<int>>& res, vector<int>& arr) {
        if (arr.size() == mx_len) {
            res.push_back(arr);
            return;
        }
        for (int idx=start; idx < nums.size(); idx++){
            arr.push_back(nums[idx]);
            backtrack_comb(idx+1, mx_len, nums, res, arr);
            arr.pop_back();
        }

    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> arr;
        for (int k=0; k < nums.size()+1; k++){
            backtrack_comb(0, k, nums, res, arr);
        }
        // for (const auto &row : res) {
        //     cout << "[ ";
        //     for (const auto &elem : row) cout << elem << " ";
        //     cout << "], ";
        // }
        return res;
    }
};